src/libstd/panicking.rs

```diff
@@ -28,11 +30,131 @@ thread_local! {
+        panic!("cannot modify the panic handler from a panicking thread");
+    }
+
+    let _lock = HANDLER_LOCK.lock();
+    unsafe {
+        let handler = HANDLER;
+        HANDLER = &default_handler;
+        if handler == &default_handler as *const _ {
+            Box::new(default_handler)
+        } else {
+            Box::from_raw(handler as *mut _)
+        }
+    }
+}
+
+///
+#[unstable(feature = "panic_handler", reason = "awaiting feedback", issue = "30449")]
+#[cfg(stage0)]
+pub fn take_handler() -> Box<Fn(&PanicInfo) + 'static + Sync + Send> {
+    panic!()
+}
+
+/// Registers a custom panic handler, replacing any that was previously
+/// registered.
+///
+/// # Panics
+///
+/// Panics if called from a panicking thread.
+#[unstable(feature = "panic_handler", reason = "awaiting feedback", issue = "30449")]
+#[cfg(not(stage0))] // Waiting on #28270 to hit snapshot
+pub fn set_handler<F>(handler: F) where F: Fn(&PanicInfo) + 'static + Sync + Send {
+    if thread::panicking() {
+        panic!("cannot modify the panic handler from a panicking thread");
+    }
+
+    let _lock = HANDLER_LOCK.lock();
+    unsafe {
+        let old_handler = HANDLER;
+        HANDLER = Box::into_raw(Box::new(handler)) as *const _;
+
+        if old_handler != &default_handler as *const _ {
+            Box::from_raw(old_handler as *mut (Fn(&PanicInfo) + 'static + Sync + Send));
+        }
+    }
+}
+
+///
+#[unstable(feature = "panic_handler", reason = "awaiting feedback", issue = "30449")]
+#[cfg(stage0)] // Waiting on #28270 to hit snapshot
+pub fn set_handler<F>(handler: F) where F: Fn(&PanicInfo) + 'static + Sync + Send {
+    panic!()
+}
+
+/// A struct providing information about a panic.
+#[unstable(feature = "panic_handler", reason = "awaiting feedback", issue = "30449")]
+pub struct PanicInfo<'a> {
+    payload: &'a (Any + Send),
+    location: &'a Location<'a>,
+}
+
+impl<'a> PanicInfo<'a> {
+    /// Returns the payload associated with the panic.
+    ///
+    /// This will commonly, but not always, be a `&'static str` or `String`.
+    #[unstable(feature = "panic_handler", reason = "awaiting feedback", issue = "30449")]
+    pub fn payload(&self) -> &(Any + Send) {
+        self.payload
+    }
+
+    /// Returns information about the location from which the panic originated,
+    /// if available.
+    #[unstable(feature = "panic_handler", reason = "awaiting feedback", issue = "30449")]
+    pub fn location(&self) -> Option<&Location> {
+        Some(self.location)
+    }
+}
+
+/// A struct containing information about the location of a panic.
+#[unstable(feature = "panic_handler", reason = "awaiting feedback", issue = "30449")]
+pub struct Location<'a> {
+    file: &'a str,
+    line: u32,
+}
+
+impl<'a> Location<'a> {
+    /// Returns the name of the source file from which the panic originated.
+    #[unstable(feature = "panic_handler", reason = "awaiting feedback", issue = "30449")]
+    pub fn file(&self) -> &str {
```

Comment:
In theory this lifetime could be lifted up to `'a`, but it may be best to stick to this which is conservativ